config/validated_config_db_connector.py

```diff
@@ -17,33 +19,79 @@ def __getattr__(self, name):
                 return self.validated_set_entry
             if name == "delete_table":
                 return self.validated_delete_table
+            if name == "mod_entry":
+                return self.validated_mod_entry
         return self.connector.__getattribute__(name)
 
     def make_path_value_jsonpatch_compatible(self, table, key, value):
+        def stringify_value():
+            nonlocal value
+            if isinstance(value, dict):
+                value = {str(k):str(v) for k, v in value.items()}
+            else:
+                value = str(value)
         if type(key) == tuple:
             path = JsonPointer.from_parts([table, '|'.join(key)]).path
+        elif type(key) == list:
+            path = JsonPointer.from_parts([table, *key]).path
         else:
             path = JsonPointer.from_parts([table, key]).path
         if value == {"NULL" : "NULL"}:
             value = {}
+        else:
+            stringify_value()
         return path, value
 
-    def create_gcu_patch(self, op, table, key=None, value=None):
-        if key:
-            path, value = self.make_path_value_jsonpatch_compatible(table, key, value)
-        else: 
-            path = "/{}".format(table)
-
+    def create_gcu_patch(self, op, table, key=None, value=None, mod_entry=False):
         gcu_json_input = []
-        gcu_json = {"op": "{}".format(op),
-                    "path": "{}".format(path)}
-        if op == "add":
-            gcu_json["value"] = value
+        if op == "add" and not self.get_table(table):
+            gcu_json = {"op": "{}".format(op),
+                        "path": "/{}".format(table),
+                        "value": {}}
+            gcu_json_input.append(gcu_json)
+
+        if op == "add" and not self.get_entry(table, key):
+            path = JsonPointer.from_parts([table, key]).path
+            gcu_json = {"op": "{}".format(op),
+                        "path": "{}".format(path),
+                        "value": {}}
+            gcu_json_input.append(gcu_json)
+         
+        def add_patch_entry():
+            if key: 
+                patch_path, patch_value = self.make_path_value_jsonpatch_compatible(table, key, value)
+            else:  
+                patch_path = "/{}".format(table)
+      
+            gcu_json = {"op": "{}".format(op),
+                        "path": "{}".format(patch_path)}
+            if op == "add":
+                gcu_json["value"] = patch_value
+
+            gcu_json_input.append(gcu_json)
+        
+        if mod_entry:
+            key_start = key
+            value_copy = copy.deepcopy(value)
+            for key_end, cleaned_value in value_copy.items():
+                key = [key_start, key_end]
+                value = cleaned_value
+                add_patch_entry()
+        else:
+            add_patch_entry()
```

Comment:
>add_patch_entry [](http://example.com/codeflow?start=12&length=15)

For set_entry, is it directly translated to jsonpatch `replace` op?

Reply:
set_entry and mod_entry both use jsonpatch `add` op. 

The difference between set_entry and mod_entry is that set_entry will remove extra fields in the db which are not in the data. So this difference is captured by differing paths in the jsonpatch. The path for mod_entry is more granular.

Reply:
set_entry is not able to be directly translated to jsonpatch `replace` op because `replace` does not allow for addition of a field that does not already exist. 

For example, if we try to add a new portchannel, we will get the following error:
```
admin@vlab-01:~$ sudo config apply-patch patch.json
Patch Applier: Patch application starting.
Patch Applier: Patch: [{"op": "replace", "path": "/PORTCHANNEL/PortChannel04", "value": {"admin_status": "up", "mtu": "9100", "lacp_key": "auto", "min_links": "1"}}]
Patch Applier: Getting current config db.
Patch Applier: Simulating the target full config after applying the patch.
Failed to apply patch
Usage: config apply-patch [OPTIONS] PATCH_FILE_PATH
Try "config apply-patch -h" for help.



Error: can't replace a non-existent object 'PortChannel04'
```